### PR TITLE
[dynamo] skip_guard_eval_unsafe stance for power users

### PR DESCRIPTION
**Pull request by anijain2305**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #140251
* #140250
* #140223

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames

### CODE REVIEW

**Review by williamwen42 — COMMENTED** [review]
Do we expect the string stance to be "default" when skip_guard_eval_unsafe=True?

File: torch/_dynamo/decorators.py [hunk 1]
```diff
@@ -97,11 +97,13 @@ class set_stance(_DecoratorContextManager):
 
     _dynamo_forbidden = True
 
-    def __init__(self, stance: str, force_backend=None) -> None:
+    def __init__(
+        self, stance: str, skip_guard_eval_unsafe: bool = False, force_backend=None
```
Comment (jansel) [inline torch/_dynamo/decorators.py:101]: Let's force these to be kwargs:
```suggestion
        self, stance: str, *, skip_guard_eval_unsafe: bool = False, force_backend=None
```

File: torch/compiler/__init__.py [hunk 2]
```diff
@@ -230,7 +230,7 @@ def disable(fn=None, recursive=True):
     return torch._dynamo.disable(fn, recursive)
 
 
-def set_stance(stance: str, force_backend=None):
+def set_stance(stance: str, skip_guard_eval_unsafe=False, force_backend=None):
```
Comment (jansel) [inline torch/compiler/__init__.py:233]:
```suggestion
def set_stance(stance: str, *, skip_guard_eval_unsafe=False, force_backend=None):
```

File: torch/csrc/dynamo/extra_state.cpp [hunk 3]
```diff
@@ -153,8 +155,18 @@ void lookup(
...
     if (valid) {
       try {
-        valid = torch::dynamo::run_root_guard_manager(
-            cache_entry.root_mgr, f_locals);
+        if (is_skip_guard_eval_unsafe) {
+          // We cant stash the diff_guard_root_mgr in the cache_entry because we
+          // create a new one on every recompilation.
+          void* diff_guard_root_mgr =
+              torch::dynamo::convert_to_root_guard_manager(
+                  cache_entry.guard_manager.attr("diff_guard_root"));
```
Comment (jansel) [inline torch/csrc/dynamo/extra_state.cpp:163]: How expensive is this?  Could we do this without doing a Python getattr (with refcountin, etc)?

**Review by jansel — CHANGES_REQUESTED** [review]

**Comment by jansel** [comment]
Maybe we should also give `stance="default"` default arg, so people can choose to only set skip_guard_eval_unsafe/force_backend

File: torch/csrc/dynamo/extra_state.cpp (hunk 3 above)
Comment (anijain2305) [reply torch/csrc/dynamo/extra_state.cpp]: Removed this. And added a new member to the cache_entry, which we update in lockstep with the python diff_guard_root update.

**Review by jansel — APPROVED** [review]

**Pull request closed without merging**